lib/use-suggestions.js

```diff
@@ -0,0 +1,104 @@
+import {
+	useCallback, useEffect, useMemo, useReducer
+} from 'haunted';
+import {
+	action, reduce
+} from '@neovici/cosmoz-utils/lib/reduce';
+import { useKeyboard } from './use-keyboard';
+
+const initial = {},
+	init = action('INIT', resolve => length => resolve(length)),
+	up = action('UP'),
+	down = action('DOWN'),
+	highlight = action('HIGHLIGHT', resolve => index => resolve(index)),
+	range = action(
+		'RANGE',
+		resolve => ({
+			firstVisible: first, lastVisible: last
+		}) =>
+			resolve({
+				first,
+				last
+			})
+	),
+	reducer = reduce(initial, handle => [
+		handle(init, (_, { payload: length }) => ({
+			length
+		})),
+		handle(down, ({
+			index: i = -1, ...state
+		}) => {
+			const {
+					length, range
+				} = state,
+				index = i < length - 1 ? i + 1 : 0;
+			return {
+				...state,
+				index,
+				scrollToIndex:
+					range && (index < range.first || index >= range.last)
+						? { index }
+						: undefined
+			};
+		}),
```

Comment:
I would really appreciate if we can be more expressive by thinking more about semantics and using helper functions to indicate what we're trying to do.

Also I think the (over)usage of the latest JS features makes it slow to read and harder to understand.

I would prefer if we could rewrite things like this something like:

```suggestion
const outOfRange = (index, range) => {
		if (range == null) {
			return;
		}
		return index < range.first || index >= range.last;
	},
	stepDown = (position = -1, maxPosition) => {
		return position < maxPosition ? position + 1  : 0;
	};
	
	// ...

		handle(down, state => ({
			...state,
			index: stepDown(state.index, state.length - 1),
			scrollToIndex: outOfRange(index, state.range) ? { index } : undefined
		}),
```

For me, each helper and also the handler behavior are a lot faster to read and understand.
Especially now with the switch to new technologies and approaches, I think it's extra important to make extra efforts for clarity.